### PR TITLE
Added missing documentation to replace PLACEHOLDERs.

### PR DESCRIPTION
**Pull request by dlibenzi**
@mruberry The `DataParallel.__call__()` did not have a PLACEHOLDER but I added documentation as well, and this need to be public.

### CODE REVIEW

File: torch_xla/core/xla_model.py [hunk 1]
```diff
@@ -30,7 +30,15 @@ def parse_xla_device(device):
 
 
 def get_xla_supported_devices(devkind=None, max_devices=None):
-  """get_xla_supported_devices (PLACEHOLDER)"""
+  """Returns a list of supported devices of a given kind.
+
+  Args:
+    devkind: If specified, one of `TPU`, `GPU` or `CPU`.
```
Comment (ailzhang) [inline torch_xla/core/xla_model.py]: This indicates we support GPU as well :P Maybe we can add a warning about GPU backend?

File: torch_xla/core/xla_model.py [hunk 2]
```diff
@@ -30,7 +30,15 @@ def parse_xla_device(device):
...
+    max_devices: The maximum number of devices to be returned of that kind.
```
Comment (ailzhang) [inline torch_xla/core/xla_model.py]: Btw a minor suggestion: I usually find it useful in PyTorch docs that we indicates what's the type of input arguments and return type. 
For example: 
```
dim (int)
dim (int, optional): Default is None
dims( list[int] or int)
name( string):
```
I find it useful myself that I can see what to pass in in 1s without reading the doc string :P  More examples can be found in https://github.com/pytorch/pytorch/blob/master/torch/_tensor_docs.py#L14.

File: torch_xla/core/xla_model.py [hunk 3]
```diff
@@ -377,7 +418,21 @@ def mark_step():
...
 
 def optimizer_step(optimizer, barrier=False, optimizer_args={}):
-  """optimizer_step (PLACEHOLDER)"""
+  """Run the provided optimizer step and issue the XLA device step computation.
+
+  Args:
+    optimizer: The `torch.Optimizer` instance whose `step()` function needs to
+      be called. The `step()` function will be called with the `optimizer_args`
+      named arguments.
+    barrier: Whether teh XLA tensor barrier should be issued in this API. If
```
Comment (zcain117) [inline torch_xla/core/xla_model.py]: teh -> the

File: torch_xla/core/xla_model.py [hunk 4]
```diff
@@ -377,7 +418,21 @@ def mark_step():
...
+      using the PyTorch XLA `ParallelLoader` or `DataParallel` support, this is
+      not necessary as the barrier will be issued by the XLA data loader
+      iterator `next()` call.
+    optimizer_args: Named arguments dictionary for the `optimizer.step()` call.
+
+  Returns:
+    the same value returned by the `optimizer.step()` call.
```
Comment (zcain117) [inline torch_xla/core/xla_model.py]: the -> The

File: torch_xla/distributed/data_parallel.py [hunk 5]
```diff
@@ -34,13 +34,19 @@ def __init__(self):
 
 
 class DataParallel(object):
-  """DataParallel (PLACEHOLDER)"""
+  """Enable to execution of a model network in replicated mode using threads.
```
Comment (zcain117) [inline torch_xla/distributed/data_parallel.py]: IMO:
Enable to execution -> Enable the execution

File: torch_xla/utils/utils.py [hunk 6]
```diff
@@ -25,7 +25,14 @@ def __init__(self):
 
 
 class SampleGenerator(object):
-  """SampleGenerator (PLACEHOLDER)"""
+  """Iterator which returns multiple samples of a given input data.
+
+  Can be used in place of a PyTorch `DataLoader` to generate synthetic data.
+
+  Args:
+    data: The input data which should be returned.
```
Comment (zcain117) [inline torch_xla/utils/utils.py]: This one is a bit vague- I think data should be a sample tensor and this method will return tensors of the same shape and dtype filled in with random values.

Could be good to give an example like:
(torch.zeros(FLAGS.batch_size, 3, img_dim, img_dim),
 torch.zeros(FLAGS.batch_size, dtype=torch.int64))

Or we could point to a sample usage:
https://github.com/pytorch/xla/blob/master/test/test_train_imagenet.py#L107

File: torch_xla/distributed/parallel_loader.py [hunk 7]
```diff
@@ -40,7 +40,22 @@ def next(self):
 
 
 class ParallelLoader(object):
-  """ParallelLoader (PLACEHOLDER)"""
+  """Wraps an existing PyTorch DataLoader with background data upload.
+
+  Args:
+    loader: The PyTorch DataLoader to be wrapped.
+    devices: The list of devices where the data has to be sent. The i-th sample
```
Comment (zcain117) [inline torch_xla/distributed/parallel_loader.py]: +1 to Ailing's suggestion about specifying variable types. For example, I think `devices` are a bit tough in this API without types. They appear sometimes as objects (like in `xla_device`), sometimes as device IDs (like in `DataParallel`), sometimes are strings (like in `get_xla_supported_devices`). I think it's fine that they appear in different ways but it's best to be more explicit in the docstring about the variable type

File: torch_xla/utils/utils.py (hunk 6 above)
Comment (dlibenzi) [reply torch_xla/utils/utils.py]: They will not be returned with random values ... but the very same exact data will be returned num_samples times.

Comment (dlibenzi) [reply torch_xla/utils/utils.py]: Links to code tend to be broken if code changes.
I tried to add a literal block (https://devguide.python.org/documenting/#source-code) but google-pyformat makes a mess out of it.

File: torch_xla/distributed/parallel_loader.py (hunk 7 above)
Comment (dlibenzi) [reply torch_xla/distributed/parallel_loader.py]: Yeah ... IMHO we should standardize on `torch.device` for this, but that requires changes on all the test models, so will be matter for a following PR.

File: torch_xla/distributed/data_parallel.py [hunk 8]
```diff
@@ -92,7 +100,29 @@ def _module_runner(self, loop_fn, device, module, loader, context, result):
       result.result = e
       self._handle_runner_exception(device, e)
 
-  def __call__(self, loop_fn, loader, fixed_batch_size=False):
+  def __call__(self, loop_fn, loader, fixed_batch_size=False, batchdim=0):
```
Comment (mruberry) [inline torch_xla/distributed/data_parallel.py]: This needs to be added to index.rst.

**Pull request merged**